File: data/code_pairs/cpp/980.cpp
#include <iostream>
#include <fstream>
#include <cstdlib>
#include <string>
#include <cmath>

// Function Declarations
void warn(const std::string& msg);
void tnoua(const std::string& ia, int nc);
void dasign(int lu, int mode, const std::string& ia, int len);
void vexit(int ierr);

int main() {
    warn("Test warn message");
    tnoua("TestTnoua", 8);
    dasign(1, -1, "example.txt", 100);
    // vexit(0); // This will terminate the program; uncomment to test

    return 0;
}

// Function Implementations
void warn(const std::string& msg) {
    std::cout << " " << msg << std::endl;
}

void tnoua(const std::string& ia, int nc) {
    std::cout << ia.substr(0, nc) << std::flush; // Using std::flush to avoid adding a newline
}

void dasign(int lu, int mode, const std::string& ia, int len) {
    // Simplified implementation example for demonstrating file operations
    std::fstream file;
    std::string filename = ia; // Assuming ia is the filename
    // Open file based on mode (This example doesn't fully replicate the Fortran logic)
    if (mode > 0) {
        file.open(filename, std::ios::out); // Simplified example
    } else {
        file.open(filename, std::ios::in); // Simplified example
    }
    if (!file.is_open()) {
        std::cerr << "Failed to open file: " << filename << std::endl;
    }
    // Perform operations with the file here
    file.close();
}

void vexit(int ierr) {
    if (ierr != 0) {
        std::cout << "exit called with error code " << ierr << std::endl;
    }
    std::exit(ierr);
}

File: data/code_pairs/fortran/980.f90
program main
    call warn('Test warn message')
    call tnoua('TestTnoua', 8)
    call dasign(1, -1, 'example.txt', 100)
    ! call vexit(0) ! This will stop the program; uncomment to test
end program main

subroutine warn(msg)
    character(*), intent(in) :: msg
    print '(A)', ' ' // trim(msg)
end subroutine warn

subroutine tnoua(ia, nc)
    character(*), intent(in) :: ia
    integer, intent(in) :: nc
    print '(A$)', ia(1:nc)
end subroutine tnoua

subroutine dasign(lu, mode, ia, len)
    integer, intent(in) :: lu, mode, len
    character(*), intent(in) :: ia
    ! Placeholder for actual file operation logic
    print *, 'dasign called with lu=', lu, ', mode=', mode, ', ia=', ia, ', len=', len
end subroutine dasign

subroutine vexit(ierr)
    integer, intent(in) :: ierr
    if (ierr /= 0) print *, 'exit called with error code', ierr
    stop
end subroutine vexit